composeml/label_maker.py

```diff
@@ -206,41 +207,141 @@ def slice(self, df, num_examples_per_instance, minimum_data=None, gap=None, drop
-
-                if df.context.slice_number >= num_examples_per_instance:
+                if ds.context.slice_number >= num_examples_per_instance:
                     break
 
+    @property
+    def _bar_format(self):
+        """Template to format the progress bar during a label search."""
+        value = "Elapsed: {elapsed} | "
+        value += "Remaining: {remaining} | "
+        value += "Progress: {l_bar}{bar}| "
+        value += self.target_entity + ": {n}/{total} "
+        return value
+
+    def _run_search(self, df, search, gap=None, min_data=None, drop_empty=True, verbose=True, *args, **kwargs):
+        """Search implementation to make label records.
+
+        Args:
+            df (DataFrame): Data frame to search and extract labels.
+            search (LabelSearch or ExampleSearch): The type of search to be done.
+            min_data (str): Minimum data before starting search. Default value is first time of index.
+            gap (str or int): Time between examples. Default value is window size.
+                If an integer, search will start on the first event after the minimum data.
+            drop_empty (bool): Whether to drop empty slices. Default value is True.
+            verbose (bool): Whether to render progress bar. Default value is True.
+            *args: Positional arguments for labeling function.
+            **kwargs: Keyword arguments for labeling function.
+
+        Returns:
+            records (list(dict)): Label Records
+        """
+        entity_groups = self.set_index(df).groupby(self.target_entity)
+        multiplier = search.expected_count if search.is_finite else 1
+        total = entity_groups.ngroups * multiplier
+
+        progress_bar, records = tqdm(
+            total=total,
+            bar_format=self._bar_format,
+            disable=not verbose,
+            file=stdout,
+        ), []
+
+        def missing_examples(entity_count):
+            return entity_count * search.expected_count - progress_bar.n
+
+        for entity_count, group in enumerate(entity_groups):
+            entity_id, df = group
+
+            slices = self._slice(
+                df=df,
+                gap=gap,
+                min_data=min_data,
+                drop_empty=drop_empty,
+            )
+
+            for ds in slices:
+                items = self.labeling_function.items()
+                labels = {name: lf(ds, *args, **kwargs) for name, lf in items}
+                valid_labels = search.is_valid_labels(labels)
+                if not valid_labels: continue
+
+                records.append({
+                    self.target_entity: entity_id,
+                    'cutoff_time': ds.context.window[0],
+                    **labels,
+                })
+
+                search.update_count(labels)
+                if search.is_finite: progress_bar.update(n=1)
+                if search.is_complete: break
+
+            n = missing_examples(entity_count + 1) if search.is_finite else 1
```

Comment:
can we add some short comments summarizing these progress bar steps?

Reply:
Yes, I added some comments about the progress bar steps.